rangefeed: remove dedicated goroutine from registration

aliher1911 opened the pull request:
Previously each rangefeed registration was creating a separate goroutine to handle outgoing events. With mux rangefeeds, multiple registrations were contending for grpc stream writer unnecessarily increasing go scheduler load.
This commit replaces goroutine with scheduler driven by mux feed itself handling all attached registrations from a single goroutine.

Release note: None

aliher1911 committed:
rangefeed: introduce processor interface

This PR extracts externally visible processor functions into interface.
This will allow us to provide alternative implementations while keeping
an option to fall back to old one.

Release note: None

aliher1911 committed:
rangefeed: add scheduler based rangefeed processor

Previously each rangefeed processor run a dedicated goroutine to
process range events. With high replica density that produced
10K's of goroutines waking up periodically and overloading go
scheduler.
This commit adds a version of processor that uses a scheduler
with a fixed pool of workers to process all ranges on the node.

TBD: list config option that enables this behaviour.

Epic: none

Release note: None

aliher1911 committed:
rangefeed: remove dedicated goroutine from registration

Previously each rangefeed registration was creating a separate
goroutine to handle outgoing events. With mux rangefeeds, multiple
registrations were contending for grpc stream writer unnecessarily
increasing go scheduler load.
This commit replaces goroutine with scheduler driven by mux feed
itself handling all attached registrations from a single goroutine.

Epic: none

Release note: None

aliher1911 committed:
util/sched: scheduler for rangefeed usage

Scheduler adds a work scheduler that could enqueue events
to be processed by callbacks using a fixed goroutine pool.
By using scheduler, components can eliminate excessive handover
of data between goroutines in per-range cases reducing go
scheduler load and thus reducing CPU and responsiveness.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107423)
<!-- Reviewable:end -->

aliher1911 committed:
testutil: make RunValues generic

Previously run values was using interface{} and casting to provide
execution options for tests. Making function generic reduces boilerplate
code needed to setup and use test parameters.

Epic: none

Release note: None
